hookspec: add pathlib.Path alternatives to py.path.local parameters in hooks

bluetech opened the pull request:
The first commit is just some scattered `py.path.local` -> `pathlib.Path` conversions.

The second commit adds `pathlib.Path` parameters to all hooks which take `py.path.local`, to allow hook implementers to convert to pathlib. Along with @RonnyPfannschmidt's work on the nodes (#8037) this should enable pytest itself to entirely move away from py.path internally, allowing for eventual deprecation.

nicoddemus commented on src/_pytest/hookspec.py:
```diff
@@ -261,27 +262,39 @@ def pytest_collection_finish(session: "Session") -> None:
     :param py.path.local path: The path to analyze.
     :param _pytest.config.Config config: The pytest config object.
+
+    versionchanged:: 6.3.0
```
```suggestion
    .. versionchanged:: 6.3.0
```

nicoddemus commented on src/_pytest/hookspec.py:
```diff
@@ -261,27 +262,39 @@ def pytest_collection_finish(session: "Session") -> None:
+    :param pathlib.Path fspath: The path to analyze.
     :param py.path.local path: The path to collect.
+
+    versionchanged:: 6.3.0
```
```suggestion
    .. versionchanged:: 6.3.0
```

nicoddemus commented on src/_pytest/hookspec.py:
```diff
@@ -261,27 +262,39 @@ def pytest_collection_finish(session: "Session") -> None:
     :param py.path.local path: The path to collect.
+
+    versionchanged:: 6.3.0
+        The ``fspath`` parameter was added as a pathlib.Path equivalent
```
I _think_ this works:

```suggestion
        The ``fspath`` parameter was added as a :class:`pathlib.Path` equivalent
```

nicoddemus commented on src/_pytest/monkeypatch.py:
```diff
@@ -325,18 +324,15 @@ def syspath_prepend(self, path) -> None:
         """
         if self._cwd is None:
             self._cwd = os.getcwd()
         if hasattr(path, "chdir"):
```
We might replace this with if/else with:

```python
Path(path).chdir()
```

bluetech replied:
Looks like `Path` doesn't have a `chdir()` method. So I just went with `os.chdir(path)` instead (i.e. removed the hasattr stuff). If something out there somehow needed that, I guess we'd find out :)

nicoddemus replied:
Oh I could swear it did. 😁 

Yeah I think `os.chdir(path)` there is the right call. 👍

nicoddemus reviewed (commented):
Cool, left a few minor comments. 👍

RonnyPfannschmidt reviewed (commented):
I just skimmed this on mobil and love what I see

We might want to add a warning for hooks using the legacy local path parameter

bluetech commented:
@nicoddemus Thanks, I addressed your comments.

@RonnyPfannschmidt My preference is to only start deprecating the `py.path` stuff once pytest itself is fully migrated, so users can do it all at once.

nicoddemus reviewed (approved):
Great work!
